### ISSUE TITLE
Add option to delete oldest snapshot "before" uploading the latest.

### ISSUE THREAD

**Issue by riddledaxis** [sabeechen/hassio-google-drive-backup#402]
I am having a problem with automatically removing the oldest HA backup in Google Drive, and syncing the latest backup.

I originally configured the add-on to keep 8  snapshots in Google Drive and 8 in HA. Until recently this was working fine. 

Until I ran out of space in Google Drive as my backups had grown to quite a large size. As I didn't want to reduce the content of my snapshots, I reduced the number being kept in Google Drive ending up at 4.

Since then the add-on will not automatically delete the oldest snapshot to release enough space to upload the latest, and I am having to do this manually through the add-on. So it is clearly not a space issue even though the add-on really believes it is and I have the required 4 snapshots in Google Drive.

![image](https://user-images.githubusercontent.com/47009467/117977221-8072c100-b328-11eb-8948-86662f21a82c.png)

These are the logs from last night's sync
```
05-12 05:43:58 DEBUG [backup.worker.watcher] Backup directory changed
05-12 05:43:58 DEBUG [backup.worker.watcher] Backup directory modified event
05-12 05:43:59 INFO [backup.ha.hasource] Snapshot finished
05-12 05:43:59 DEBUG [backup.model.syncer] Sync requested by HomeAssistant
05-12 05:43:59 INFO [backup.model.coordinator] Syncing Snapshots
05-12 05:43:59 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
05-12 05:43:59 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots/3e03d25d/info
05-12 05:43:59 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/?q=%2717IAxNYVGFv1X5A51WPtG4XWqZ_xIeNk6%27+in+parents&fields=nextPageToken%2Cfiles%28id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents%29&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true&corpora=allDrives
05-12 05:43:59 INFO [backup.drive.drivesource] Uploading 'Full Snapshot 2021.5.2' to Google Drive
05-12 05:43:59 DEBUG [backup.drive.driverequests] Starting a new upload session with Google Drive
05-12 05:43:59 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true
05-12 05:44:00 ERROR [backup.model.coordinator] Google Drive is out of space
05-12 05:44:00 INFO [backup.model.coordinator] I'll try again in an hour
05-12 05:44:03 INFO [backup.worker.watcher] Backup directory changed
05-12 05:44:03 DEBUG [backup.model.syncer] Sync requested by Backup Directory Watcher
05-12 05:44:03 INFO [backup.model.coordinator] Syncing Snapshots
05-12 05:44:03 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
05-12 05:44:03 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/?q=%2717IAxNYVGFv1X5A51WPtG4XWqZ_xIeNk6%27+in+parents&fields=nextPageToken%2Cfiles%28id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents%29&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true&corpora=allDrives
05-12 05:44:04 INFO [backup.drive.drivesource] Uploading 'Full Snapshot 2021.5.2' to Google Drive
05-12 05:44:04 DEBUG [backup.drive.driverequests] Starting a new upload session with Google Drive
05-12 05:44:04 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true
05-12 05:44:04 ERROR [backup.model.coordinator] Google Drive is out of space
05-12 05:44:04 INFO [backup.model.coordinator] I'll try again in an hour
05-12 05:44:06 INFO [backup.debugworker] Sending error report (see settings to disable)
```
I am running HassOS 5.13 - core-2021.5.2 - supervisor-20121.04.03 and version 0.102.0 of the add-on

This is my add-on config
![image](https://user-images.githubusercontent.com/47009467/117977170-73ee6880-b328-11eb-9c68-44da756d17a5.png)

Please let me know what I need to change, as it is becoming a pain to have to manually delete a snapshot through the add-on.

Thank you

**Comment by sabeechen** [comment]
The addon doesn't delete the oldest backup until the newest backup has been uploaded, which is [by design](https://github.com/sabeechen/hassio-google-drive-backup#i-configured-this-to-only-keep-4-snapshots-in-drive-and-home-assistant-but-sometimes-i-can-see-there-are-5).   I will likely be adding an option to change this behavior in the future, but for now if you configure the addon to store N snapshots, then it needs space for N+1 to function properly.

I might suggest that you avoid running the addon near its storage limit to make it more reliable, as small changes in the size of what your snapshots can lead to failure.
* Consider using partial snapshots.  Most people backup the full sensor database, media files, camera recordings, etc.  Unless you have a specific use case, these items are typically nice to to have in the event of a failure but not particularly critical.
* Google Drive offers 200GB for $3/mo, which isn't that much in the grand scheme of things.  But it is real money.

**Comment by sabeechen** [comment]
It took a bit for me to get through everything in the latest release, but I've added this in the just released v0.105.1.  Look for it in the setting under "Delete the oldest backup before creating a new one"
You can wait for HA to notice the new version, or go to "Supervisor" > "Addon Store" > "..." > "Reload"  to force a check.
Please feel free to re-open this issue if you think there is something else I can help with.
